pkg/controller/revert.go

```diff
@@ -12,25 +12,56 @@ import (
 )
 
 func (c *Controller) Revert(name string) error {
-	doable := false
 	rw := false
 	wo := false
 	for _, rep := range c.replicas {
-		//BUG: Need to deal with replica in rebuilding(WO) mode.
+		// BUG: Need to deal with replica in rebuilding(WO) mode.
 		if rep.Mode == types.RW {
 			rw = true
 		} else if rep.Mode == types.WO {
 			wo = true
 		}
 	}
-	if rw && !wo {
-		doable = true
-	}
-
+	doable := rw && !wo
 	if !doable {
 		return fmt.Errorf("not valid state to revert, rebuilding in process or all replica are in ERR state")
 	}
 
+	snapshotDiskName := GenerateSnapshotDiskName(name)
+	found := false
+	for _, r := range c.replicas {
+		if r.Mode != types.RW {
+			continue
+		}
+
+		disks, _, err := GetReplicaDisksAndHead(r.Address)
+		if err != nil {
+			return err
+		}
+		for _, disk := range disks {
+			if disk.Name != snapshotDiskName {
+				continue
+			}
+			if disk.Removed {
+				return fmt.Errorf("cannot revert to removed snapshot %v", name)
+			}
+			// for c := range disk.Children {
+			// 	if c == head {
+			// 		logrus.Infof("there is no need to do revert since the parent of the current volume head is already snapshot %v", name)
+			// 		return nil
+			// 	}
+			// }
```

Comment:
Should we check that the revert target is the snapshot that is directly behind the volume head?

Reply:
If the current volume head has data then the volume is reverted to the previous snapshot, we should still create a new empty volume head pointing to the snapshot, right? If yes, we don't need this check.

Reply:
YES. I just tested this part. The volume will discard the old volume head regardless of the reverting snapshot being the parent.
I have removed this code.